iceoryx_posh/include/iceoryx_posh/popo/sample.hpp

```diff
@@ -35,40 +71,57 @@ template <typename T>
 class Sample
 {
   public:
+    /// @brief constructor for a Sample used by the Publisher
+    /// @tparam S is a dummy template parameter to enable the constructor only for non-const T
+    /// @param samplePtr is a `rvalue` to a `cxx::unique_ptr<T>` with to the data of the encapsulated type T
+    /// @param publisher is a reference to the publisher to be able to use the `publish` and `release` methods
+    template <typename S = T, typename = std::enable_if_t<std::is_same<S, T>::value && !std::is_const<S>::value, S>>
     Sample(cxx::unique_ptr<T>&& samplePtr, PublisherInterface<T>& publisher);
-    Sample(const Sample<T>&) = delete;
-    Sample<T>& operator=(const Sample<T>&) = delete;
+
+    /// @brief constructor for a Sample used by the Subscriber
+    /// @tparam S is a dummy template parameter to enable the constructor only for const T
+    /// @param samplePtr is a `rvalue` to a `cxx::unique_ptr<T>` with to the data of the encapsulated type T
+    template <typename S = T, typename = std::enable_if_t<std::is_same<S, T>::value && std::is_const<S>::value, S>>
+    Sample(cxx::unique_ptr<T>&& samplePtr) noexcept;
+
+    Sample(std::nullptr_t) noexcept;
+    ~Sample();
+
     Sample<T>& operator=(Sample<T>&& rhs);
     Sample(Sample<T>&& rhs);
-    ~Sample();
 
-    Sample(std::nullptr_t) noexcept;
+    Sample(const Sample<T>&) = delete;
+    Sample<T>& operator=(const Sample<T>&) = delete;
+
 
     ///
     /// @brief operator -> Transparent access to the encapsulated type.
-    /// @return
+    /// @return a pointer to the encapsulated type.
+    /// @details Only available for non-const type T.
     ///
+    template <typename S = T, typename = std::enable_if_t<std::is_same<S, T>::value && !std::is_const<S>::value, S>>
     T* operator->() noexcept;
 
     ///
     /// @brief operator -> Transparent read-only access to the encapsulated type.
-    /// @return
+    /// @return a const pointer to the encapsulated type.
     ///
     const T* operator->() const noexcept;
 
     ///
     /// @brief operator* Provide a reference to the encapsulated type.
     /// @return A T& to the encapsulated type.
-    /// @details Only available for non void type T.
+    /// @details Only available for non-const type T.
     ///
-    template <typename S = T,
-              typename = std::enable_if_t<std::is_same<S, T>::value && !std::is_same<S, void>::value, S>>
+    template <
```

Comment:
This `enable_if` construct will cause ugly template errors. I think in this case we could also use `static_assert` and provide the user a nice to read error message.

Reply:
This is just to enable some methods only for the publisher, which can mutate the sample.